src/software/ai/navigator/navigator.cpp

```diff
@@ -12,197 +13,158 @@ Navigator::Navigator(std::unique_ptr<PathManager> path_manager,
-                intent.getMotionConstraints(), world);
-        obstacles.insert(obstacles.end(), motion_constraint_obstacles.begin(),
-                         motion_constraint_obstacles.end());
-
-        if (intent.getBallCollisionType() == BallCollisionType::AVOID)
+        auto navigator_params = intent->getNavigatorParams();
+        if (navigator_params)
         {
-            auto ball_obstacle = robot_navigation_obstacle_factory.createFromBallPosition(
-                world.ball().position());
-            obstacles.push_back(ball_obstacle);
-        }
-
-        auto robot = world.friendlyTeam().getRobotById(intent.getRobotId());
-
-        if (robot)
-        {
-            Point start = robot->position();
-            Point end   = intent.getDestination();
-
-            path_objectives.insert(PathObjective(start, end, robot->velocity().length(),
-                                                 obstacles, intent.getRobotId()));
+            if (robot_id_to_path.find(intent->getRobotId()) != robot_id_to_path.end())
+            {
+                auto path = robot_id_to_path.at(intent->getRobotId());
+                if (path)
+                {
+                    auto [destination, final_speed] = calculateDestinationAndFinalSpeed(
+                        navigator_params.value(), *path, intent, world);
+
+                    robot_primitives_map[intent->getRobotId()] =
+                        intent->getUpdatedPrimitiveMsg(destination, final_speed);
+                    continue;
+                }
+            }
+            LOG(WARNING)
+                << "Navigator's path manager could not find a path for RobotId = "
+                << intent->getRobotId();
+            robot_primitives_map[intent->getRobotId()] =
+                ProtoCreatorPrimitiveVisitor().createPrimitiveMsg(
+                    StopPrimitive(intent->getRobotId(), false));
         }
         else
         {
-            std::stringstream ss;
-            ss << "Failed to find robot associated with robot id = "
-               << intent.getRobotId();
-            LOG(WARNING) << ss.str();
+            robot_primitives_map[intent->getRobotId()] = intent->getPrimitiveMsg();
         }
     }
-    return path_objectives;
+
+    return primitive_set_msg;
 }
 
-std::vector<std::unique_ptr<Primitive>> Navigator::getPrimitivesFromMoveIntents(
-    const std::vector<MoveIntent> &move_intents, const World &world)
+std::unordered_set<PathObjective> Navigator::getPathObjectivesFromIntents(
+    const std::vector<std::unique_ptr<Intent>> &intents, const World &world,
+    const std::vector<ObstaclePtr> &friendly_non_navigating_robot_obstacles)
 {
-    std::vector<std::unique_ptr<Primitive>> primitives;
+    std::unordered_set<PathObjective> path_objectives;
 
-    Rectangle navigable_area = world.field().fieldBoundary();
+    for (const auto &intent : intents)
+    {
+        auto navigator_params = intent->getNavigatorParams();
```

Comment:
It would make sense to have a `getPathObjectiveFromIntent` function since there's nothing shared between different iterations of this loops except for `path_objectives`

Reply:
refactored, but in a different way